File: src/BIOME4Py/growth_subroutines/calcphi.py
"""Calculate the maximum quantum yield of photosynthesis."""

# Standard library
from typing import List


def calcphi(gpp: List[float]) -> float:
    totgpp = 0.0
    svar = [0.0, 0.0, 0.0, 0.0]
    snormavg = [0.0, 0.0, 0.0, 0.0]

    # Calculate total GPP and mean GPP
    totgpp = sum(gpp)
    meangpp = totgpp / 12.0

    # Normalize GPP
    normgpp = [g / meangpp for g in gpp]

    # Calculate seasonal normalized averages
    snormavg[0] = sum(normgpp[0:3]) / 3.0
    snormavg[1] = sum(normgpp[3:6]) / 3.0
    snormavg[2] = sum(normgpp[6:9]) / 3.0
    snormavg[3] = sum(normgpp[9:12]) / 3.0

    # Calculate population variances by season
    for m in range(3):
        a = ((normgpp[m] - snormavg[0]) ** 2) / 3.0
        svar[0] += a

    for m in range(3, 6):
        a = ((normgpp[m] - snormavg[1]) ** 2) / 3.0
        svar[1] += a

    for m in range(6, 9):
        a = ((normgpp[m] - snormavg[2]) ** 2) / 3.0
        svar[2] += a

    for m in range(9, 12):
        a = ((normgpp[m] - snormavg[3]) ** 2) / 3.0
        svar[3] += a

    avar = sum(svar)

    # Calculate phi based on the annual variability
    phi = 0.3518717 * avar + 0.2552359
    if phi >= 1.0:
        phi /= 10.0

    return phi


if __name__ == "__main__":
    calcphi()

File: src/models/MechanisticModel/growth_subroutines/calcphi.jl
"""
calcphi(gpp::AbstractVector{T})::T

Calculate the maximum quantum yield of photosynthesis based on GPP data.

# Arguments
- `gpp`: An array of 12 T values representing Gross Primary Productivity for each month.

# Returns
- `phi`: The calculated maximum quantum yield of photosynthesis.
"""
function calcphi(gpp::AbstractVector{T})::T where {T <: Real}
@assert length(gpp) == 12 "gpp must have exactly 12 elements"

totgpp = sum(gpp)
meangpp = totgpp / T(12.0)
normgpp = [g / meangpp for g in gpp]

snormavg = T[0.0, 0.0, 0.0, 0.0]
snormavg[1] = sum(normgpp[1:3]) / T(3.0)
snormavg[2] = sum(normgpp[4:6]) / T(3.0)
snormavg[3] = sum(normgpp[7:9]) / T(3.0)
snormavg[4] = sum(normgpp[10:12]) / T(3.0)

svar = T[0.0, 0.0, 0.0, 0.0]
for m in 1:3
    a = ((normgpp[m] - snormavg[1]) ^ 2) / T(3.0)
    svar[1] += a
end

for m in 4:6
    a = ((normgpp[m] - snormavg[2]) ^ 2) / T(3.0)
    svar[2] += a
end

for m in 7:9
    a = ((normgpp[m] - snormavg[3]) ^ 2) / T(3.0)
    svar[3] += a
end

for m in 10:12
    a = ((normgpp[m] - snormavg[4]) ^ 2) / T(3.0)
    svar[4] += a
end

avar = sum(svar)
phi = T(0.3518717) * avar + T(0.2552359)

if phi >= T(1.0)
    phi /= T(10.0)
end

return phi
end
